src/lib/RoutingManager.js

```diff
@@ -0,0 +1,195 @@
+ * @property {(object) => ()} write push a route object into a storage
+ * @property {((object) => (object)) => ()} onUpdate sets an event listenere when the storage is updated by a third party
+ * @property {() => (object)} read reads the storage and gets a route object
+ * @property {(object) => string} createURL transforms a route object into a URL
+ * @property {() => ()} dispose cleans up any event listeners
+ */
+
+export default class RoutingManager {
+  constructor({ instantSearchInstance, router, stateMapping } = {}) {
+    this.originalConfig = null;
+    this.firstRender = true;
+
+    this.router = router;
+    this.stateMapping = stateMapping;
+    this.instantSearchInstance = instantSearchInstance;
+
+    this.originalUIState = this.stateMapping.routeToState(this.router.read());
+  }
+
+  init({ state }) {
+    // store the initial state from the storage
+    // so that we can compare it with the state after the first render
+    // in case the searchFunction has modifyied it.
+    const { helper } = this.instantSearchInstance;
+    this.initState = this.getAllUIStates({
+      helper,
+      state,
+      instantSearchInstance: this.instantSearchInstance,
+    });
+  }
+
+  getConfiguration(currentConfiguration) {
+    // we need to create a REAL helper to then get its state. Because some parameters
+    // like hierarchicalFacet.rootPath are then triggering a default refinement that would
+    // be not present if it was not going trough the SearchParameters constructor
+    this.originalConfig = algoliasearchHelper(
+      { addAlgoliaAgent() {} },
+      currentConfiguration.index,
+      currentConfiguration
+    ).state;
+    return Object.assign(
+      {},
+      this.getAllSearchParameters({
+        state: this.originalConfig,
+        uiState: this.originalUIState,
+      })
+    );
+  }
+
+  render({ state }) {
+    if (this.firstRender) {
+      this.firstRender = false;
+      const { helper } = this.instantSearchInstance;
+      this.router.onUpdate(route => {
+        const uiState = this.stateMapping.routeToState(route);
+        const currentUIState = this.getAllUIStates({
+          helper,
+          state: helper.state,
+          instantSearchInstance: this.instantSearchInstance,
+        });
+
+        if (isEqual(uiState, currentUIState)) return;
+
+        // retrieve a search state from the widgets and the UI state
+        const searchParameters = this.getAllSearchParameters({
+          state,
+          instantSearchInstance: this.instantSearchInstance,
+          uiState,
+        });
+
+        const fullHelperState = Object.assign(
+          {},
+          this.originalConfig,
+          searchParameters
+        );
+
+        if (isEqual(fullHelperState, searchParameters)) return;
+
+        helper
+          .overrideStateWithoutTriggeringChangeEvent(searchParameters)
+          .search();
+      });
+
+      helper.on('change', searchState => {
+        // this.renderURLFromState(s)
```

Comment:
We forgot to move this function to `renderURLFromState`, we can't remove the listener on the helper in the `dispose` method with the current implementation.